File: internal/config/config_update_sync_test.go
package config

import (
	"os"
	"path/filepath"
	"testing"
)

// TestLoad_UpdateAndSyncNamespaces exercises AC2/AC3 of
// REQ-2026-08-02-unificar-a-leitura-do-trackfw-yaml-em-um-unico-carregador-nos-tres-clis.md: the
// eleven fields historically read by the artisanal scanners (update.go / linear.go / jira.go)
// resolve as strings via the single config.Load() path, exposed under cfg.Update and cfg.Sync.
// Keys stay flat at the YAML root — only the in-memory struct is namespaced.
func TestLoad_UpdateAndSyncNamespaces(t *testing.T) {
	Reset()
	tmp := t.TempDir()
	orig, _ := os.Getwd()
	defer func() { _ = os.Chdir(orig) }()
	if err := os.Chdir(tmp); err != nil {
		t.Fatal(err)
	}

	yaml := `hooks: husky
ci: github
backend: go
frontend: react
pkg_manager: npm
linear_api_key: lin_api_abc123
linear_team_id: TEAM-1
jira_base_url: "https://x.atlassian.net:443"
jira_email: bot@example.com
jira_token: jira_tok_xyz
jira_project: PROJ
`
	if err := os.WriteFile(filepath.Join(tmp, "trackfw.yaml"), []byte(yaml), 0644); err != nil {
		t.Fatal(err)
	}

	cfg := Load()

	if cfg.Update.Hooks != "husky" {
		t.Errorf("Update.Hooks: want husky, got %q", cfg.Update.Hooks)
	}
	if cfg.Update.CI != "github" {
		t.Errorf("Update.CI: want github, got %q", cfg.Update.CI)
	}
	if cfg.Update.Backend != "go" {
		t.Errorf("Update.Backend: want go, got %q", cfg.Update.Backend)
	}
	if cfg.Update.Frontend != "react" {
		t.Errorf("Update.Frontend: want react, got %q", cfg.Update.Frontend)
	}
	if cfg.Update.PkgManager != "npm" {
		t.Errorf("Update.PkgManager: want npm, got %q", cfg.Update.PkgManager)
	}
	if cfg.Sync.LinearAPIKey != "lin_api_abc123" {
		t.Errorf("Sync.LinearAPIKey: want lin_api_abc123, got %q", cfg.Sync.LinearAPIKey)
	}
	if cfg.Sync.LinearTeamID != "TEAM-1" {
		t.Errorf("Sync.LinearTeamID: want TEAM-1, got %q", cfg.Sync.LinearTeamID)
	}
	if cfg.Sync.JiraBaseURL != "https://x.atlassian.net:443" {
		t.Errorf("Sync.JiraBaseURL: want https://x.atlassian.net:443, got %q", cfg.Sync.JiraBaseURL)
	}
	if cfg.Sync.JiraEmail != "bot@example.com" {
		t.Errorf("Sync.JiraEmail: want bot@example.com, got %q", cfg.Sync.JiraEmail)
	}
	if cfg.Sync.JiraToken != "jira_tok_xyz" {
		t.Errorf("Sync.JiraToken: want jira_tok_xyz, got %q", cfg.Sync.JiraToken)
	}
	if cfg.Sync.JiraProject != "PROJ" {
		t.Errorf("Sync.JiraProject: want PROJ, got %q", cfg.Sync.JiraProject)
	}
}

// TestLoad_UpdateAndSyncNamespaces_AbsentDefaultsEmpty confirms the default for an absent field
// in each of the eleven keys is "" (AC per ML-1A spec: "Default de campo ausente no YAML: string
// vazia, nos 3 CLIs").
func TestLoad_UpdateAndSyncNamespaces_AbsentDefaultsEmpty(t *testing.T) {
	Reset()
	tmp := t.TempDir()
	orig, _ := os.Getwd()
	defer func() { _ = os.Chdir(orig) }()
	if err := os.Chdir(tmp); err != nil {
		t.Fatal(err)
	}

	cfg := Load()

	if cfg.Update.Hooks != "" || cfg.Update.CI != "" || cfg.Update.Backend != "" ||
		cfg.Update.Frontend != "" || cfg.Update.PkgManager != "" {
		t.Errorf("Update namespace: want all empty strings, got %+v", cfg.Update)
	}
	if cfg.Sync.LinearAPIKey != "" || cfg.Sync.LinearTeamID != "" || cfg.Sync.JiraBaseURL != "" ||
		cfg.Sync.JiraEmail != "" || cfg.Sync.JiraToken != "" || cfg.Sync.JiraProject != "" {
		t.Errorf("Sync namespace: want all empty strings, got %+v", cfg.Sync)
	}
}

File: npm/tests/config_update_sync.test.js
'use strict'
const assert = require('assert')
const fs = require('fs')
const os = require('os')
const path = require('path')
const config = require('../src/config/index.js')

function withTmpDir(yaml, fn) {
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'trackfw-config-update-sync-'))
  try {
    if (yaml) fs.writeFileSync(path.join(tmp, 'trackfw.yaml'), yaml, 'utf8')
    config.reset()
    fn(tmp)
  } finally {
    config.reset()
    fs.rmSync(tmp, { recursive: true, force: true })
  }
}

let passed = 0, failed = 0

function test(name, fn) {
  try { fn(); console.log(`✓ ${name}`); passed++ }
  catch (e) { console.error(`✗ ${name}: ${e.message}`); failed++ }
}

// AC2/AC3 da REQ-2026-08-02-unificar-a-leitura-do-trackfw-yaml-em-um-unico-carregador-nos-tres-clis:
// os onze campos historicamente lidos pelos scanners artesanais (update.js / sync.js) resolvem
// como string via o único caminho config.load(), expostos em cfg.update e cfg.sync. As chaves
// continuam planas na raiz do YAML — só o objeto em memória é namespaced.
test('update e sync — onze campos resolvidos como string', () => {
  const yaml = [
    'hooks: husky',
    'ci: github',
    'backend: go',
    'frontend: react',
    'pkg_manager: npm',
    'linear_api_key: lin_api_abc123',
    'linear_team_id: TEAM-1',
    'jira_base_url: "https://x.atlassian.net:443"',
    'jira_email: bot@example.com',
    'jira_token: jira_tok_xyz',
    'jira_project: PROJ',
    '',
  ].join('\n')
  withTmpDir(yaml, (tmp) => {
    const cfg = config.load(tmp)
    assert.strictEqual(cfg.update.hooks, 'husky')
    assert.strictEqual(cfg.update.ci, 'github')
    assert.strictEqual(cfg.update.backend, 'go')
    assert.strictEqual(cfg.update.frontend, 'react')
    assert.strictEqual(cfg.update.pkgManager, 'npm')
    assert.strictEqual(cfg.sync.linearApiKey, 'lin_api_abc123')
    assert.strictEqual(cfg.sync.linearTeamId, 'TEAM-1')
    assert.strictEqual(cfg.sync.jiraBaseUrl, 'https://x.atlassian.net:443')
    assert.strictEqual(cfg.sync.jiraEmail, 'bot@example.com')
    assert.strictEqual(cfg.sync.jiraToken, 'jira_tok_xyz')
    assert.strictEqual(cfg.sync.jiraProject, 'PROJ')
  })
})

// Default de campo ausente no YAML: string vazia, nos 3 CLIs (spec do ML-1A).
test('update e sync — campos ausentes default para string vazia', () => {
  withTmpDir(null, (tmp) => {
    const cfg = config.load(tmp)
    assert.strictEqual(cfg.update.hooks, '')
    assert.strictEqual(cfg.update.ci, '')
    assert.strictEqual(cfg.update.backend, '')
    assert.strictEqual(cfg.update.frontend, '')
    assert.strictEqual(cfg.update.pkgManager, '')
    assert.strictEqual(cfg.sync.linearApiKey, '')
    assert.strictEqual(cfg.sync.linearTeamId, '')
    assert.strictEqual(cfg.sync.jiraBaseUrl, '')
    assert.strictEqual(cfg.sync.jiraEmail, '')
    assert.strictEqual(cfg.sync.jiraToken, '')
    assert.strictEqual(cfg.sync.jiraProject, '')
  })
})

console.log(`\n${passed} passed, ${failed} failed`)
if (failed > 0) process.exit(1)
